### PR TITLE
refactor: More directory layout refactoring

### PR DESCRIPTION
**Pull request by jskeet**

### CODE REVIEW

File: tools/Google.Cloud.Tools.ReleaseManager/GenerateApisCommand.cs [hunk 1]
```diff
@@ -383,7 +382,7 @@ private int ExecuteForUnconfigured(string[] args)
     private void GenerateUnconfigured(string arg)
     {
         Console.WriteLine($"{DateTime.UtcNow:yyyy-MM-dd'T'HH:mm:ss.fff}Z Generating {arg}");
-        var apiDirectory = Path.Combine(_rootLayout.GeneratorOutput, arg);
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.ReleaseManager/GenerateApisCommand.cs:386]: This is deliberately a change btw - it was broken before. (It should have been relative to _rootLayout.Googleapis.)

**Review by amanda-tarafa — APPROVED** [review]
Approving but there's a bug on one of the snippets source properties, I've flagged it there.

Otherwise, the only thing I'm not certain about is the sepration in two of ApiLayout. I understand the need to use one or the other of repo-root and output, but other than that I don't understand why we'd need the distinction. And the using of one root or the other can be solved with factory methods. But, I know I might be missing something, so this is not a blocker for me.

File: tools/Google.Cloud.Tools.Common/DocsLayout.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    /// Creates a layout which has <see cref="SnippetSourceDirectory"/> and <see cref="SnippetOutputDirectory"/> set
+    /// appropriately for the snippets in Google.Cloud.Docs.Snippets. All other properties are null.
+    /// </summary>
+    public static DocsLayout ForHelpSnippets(RootLayout rootLayout) => new(rootLayout, helpSnippetsMarker: null);
+
+    public static DocsLayout ForApi(RootLayout rootLayout, string id) => new(rootLayout, id);
+
+    // Type just to allow us to write a constructor that is obviously geared towards help snippets,
+    // and ensure that the calling code is clear too.
+    private sealed class HelpSnippetsMarker { }
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.Common/DocsLayout.cs:80]: Yes, I've basically been trying to localize magic, as it were - being explicit in code where possible.

File: tools/Google.Cloud.Tools.Common/GeneratorApiLayout.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    /// <summary>
+    /// Directory for production source code for the library.
+    /// </summary>
+    public string ProductionDirectory => Path.Combine(SourceDirectory, Id);
+
+    /// <summary>
+    /// Directory for snippets for the library, including hand-written snippets.
+    /// </summary>
+    public string SnippetsDirectory => Path.Combine(SourceDirectory, $"{Id}.GeneratedSnippets");
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.Common/GeneratorApiLayout.cs:38]: Fixed, thanks.

File: tools/Google.Cloud.Tools.Common/GeneratorApiLayout.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    /// <summary>
+    /// Directory for purely generated snippets for the library.
+    /// </summary>
+    public string GeneratedSnippetsDirectory => Path.Combine(SourceDirectory, $"{Id}.GeneratedSnippets");
+
+    /// <summary>
+    /// Directory for purely generated snippets for the library.
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.Common/GeneratorApiLayout.cs:46]: Done.

File: tools/Google.Cloud.Tools.Common/GeneratorApiLayout.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    /// </summary>
+    public string UnitTestsDirectory => Path.Combine(SourceDirectory, $"{Id}.Tests");
+
+    /// <summary>
+    /// Directory for purely generated snippets for the library.
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.Common/GeneratorApiLayout.cs:51]: Done.

File: tools/Google.Cloud.Tools.CompareVersions/Program.cs [hunk 6]
```diff
@@ -90,8 +90,8 @@ private static Task<AssemblyDefinition> LoadAsync(CommandLineOptions options, st
             if (version == "local")
             {
                 var rootLayout = RootLayout.ForCurrentDirectory();
-                var sourceRoot = rootLayout.CreateApiLayout(options.Package).SourceDirectory;
-                file = Path.Combine(sourceRoot, options.Package, "bin", "Release", options.Framework, $"{options.Package}.dll");
+                var productionRoot = rootLayout.CreateRepositoryApiLayout(options.Package).ProductionDirectory;
+                file = Path.Combine(rootLayout.CreateRepositoryApiLayout(options.Package).GetReleaseAssembly(options.Framework));
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.CompareVersions/Program.cs:94]: Yup, it's much smaller now.

File: tools/Google.Cloud.Tools.ReleaseManager/GenerateApisCommand.cs [hunk 7]
```diff
@@ -564,6 +563,9 @@ bool HasApis(string file)
         }
     }
 
+    private string GetApiProtosDirectory(ApiMetadata api) => GetApiProtosDirectory(api.ProtoPath);
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.ReleaseManager/GenerateApisCommand.cs:566]: No, that doesn't know the path to the protos for the API (and can't, if we've just provided an ID).

**Comment by jskeet** [comment]
> Approving but there's a bug on one of the snippets source properties, I've flagged it there.
> 
> Otherwise, the only thing I'm not certain about is the sepration in two of ApiLayout. I understand the need to use one or the other of repo-root and output, but other than that I don't understand why we'd need the distinction. And the using of one root or the other can be solved with factory methods. But, I know I might be missing something, so this is not a blocker for me.

Separating the two into different types with different properties makes it less likely that someone will use the wrong one. For example, because `GeneratorApiLayout` doesn't have `CreateDocsLayout`, you can't accidentally try to use the generator layout (which is *incapable* of building docs) when you need anything about docs. In the other direction, if you're using anything about "tweaks" but you're expecting a full repo, then there's *probably* something wrong. (Obviously we can add properties if we really need to.)

**Review by amanda-tarafa — APPROVED** [review]
Yep, looks good, thanks!

**Pull request merged**